test/e2e/recipe/index.ts

```diff
@@ -0,0 +1,175 @@
+    log(`Preparing recipe '${this.description}'`);
+
+    const timeout = this._recipe.metadata.timeout || 30_000;
+    // macOS runs quite slowly in GitHub actions
+    context.timeout(process.platform === 'darwin' ? timeout * 2 : timeout);
+
+    const appPath = join(testBasePath, this.appName);
+
+    // Drop all the files
+    for (const file of Object.keys(this._recipe.files)) {
+      log(`Writing file '${file}'`);
+
+      const path = join(appPath, file);
+
+      mkdirSync(dirname(path), { recursive: true });
+      let content = this._recipe.files[file];
+
+      // Replace with the test server localhost DSN
+      content = content.replace('__DSN__', 'http://37f8a2ee37c0409d8970bc7559c7c7e4@localhost:8123/277345');
+
+      // We replace the @sentry/electron dependency in package.json with
+      // the path to the tarball
+      if (file.endsWith('package.json')) {
+        content = content.replace(
+          /"@sentry\/electron": ".*"/,
+          `"@sentry/electron": "file:./../../../../sentry-electron-${SDK_VERSION}.tgz"`,
+        );
+      }
+
+      writeFileSync(path, content);
+    }
+
+    if (this._recipe.metadata.command) {
+      log(`Running command '${this._recipe.metadata.command}'`);
+
+      const result = spawnSync(this._recipe.metadata.command, {
+        shell: true,
+        cwd: appPath,
+        stdio: process.env.DEBUG ? 'inherit' : 'pipe',
+      });
+
+      if (result.status) {
+        console.error(result.stdout?.toString());
+        console.error(result.stderr?.toString());
+        throw new Error('Recipe command failed');
+      }
+    }
+
+    return [appPath, this.appName];
+  }
+
+  public async runTests(context: TestContext, testServer: TestServer): Promise<void> {
+    await context.start();
+
+    // Main process native crashes with Sentry Uploader are sent on the next run
+    if (this._recipe.metadata.runTwice) {
+      await context.waitForAppClose();
+      await context.stop({ retainData: true });
+      await context.start({ secondRun: true });
+    }
+
+    // Filter out events that are not for this platform/version
+    const expectedEvents = this._recipe.expectedEvents.filter(
+      (event) =>
+        event.condition === undefined || evaluateCondition('event comparison', this._electronVersion, event.condition),
+    );
+
+    await context.waitForEvents(testServer, expectedEvents.length);
+
+    if (expectedEvents.length !== testServer.events.length) {
+      throw new Error(`Expected ${expectedEvents.length} events but server has ${testServer.events.length} events`);
+    }
+
+    for (const event of testServer.events) {
+      event.data = normalize(event.data);
+    }
+
+    for (const [i, expectedEvent] of expectedEvents.entries()) {
+      delete expectedEvent.condition;
+
+      log(`Comparing event ${i + 1} of ${expectedEvents.length}`);
+      expect(testServer.events).to.containSubset([expectedEvent]);
```

Comment:
I was thinking it would be useful to leverage snapshot tests here somehow (so devs can automatically regenerate them if needed), but I guess that's harder to get working if we assert on the subset.

Reply:
I considered snapshots but the json events are modified/cleaned up to make the tests more reliable across Electron versions. For example, for most of the tests, I've stripped all the breadcrumbs out because these vary depending on the events supported for each Electron version.